From 7757509a3984c6873dda0033ccb1e7fa34369a6b Mon Sep 17 00:00:00 2001
From: David McKay <david.mckay@codasip.com>
Date: Thu, 21 Mar 2024 15:16:05 +0000
Subject: [PATCH 16/16] riscv: hobgoblin: Xilinx Axi Ethernet support

This adds support on the hobgoblin platform for the Xilinx Axi Ethernet.
This is a much more capable Ethernet than then EthernetLite. It supports
DMA and TCP checksum offload.

Performance is much improved over the EthernetLite, the downside of
this design is that it does require a license for the Ethernet block.

Added in AXI interconnect clock and removed the spurious clock that was
there.

Signed-off-by: David McKay <david.mckay@codasip.com>
---
 arch/riscv/boot/dts/codasip/Makefile          |  1 +
 .../dts/codasip/a70x-hobgoblin-axi-emac.dts   | 41 +++++++++++++++++++
 .../riscv/boot/dts/codasip/a70x-hobgoblin.dts | 13 +++---
 .../configs/codasip-a70x-hobgoblin_defconfig  |  3 ++
 4 files changed, 52 insertions(+), 6 deletions(-)
 create mode 100644 arch/riscv/boot/dts/codasip/a70x-hobgoblin-axi-emac.dts

diff --git a/arch/riscv/boot/dts/codasip/Makefile b/arch/riscv/boot/dts/codasip/Makefile
index 6df53f663347..e32a77d5f671 100644
--- a/arch/riscv/boot/dts/codasip/Makefile
+++ b/arch/riscv/boot/dts/codasip/Makefile
@@ -1,3 +1,4 @@
 # SPDX-License-Identifier: GPL-2.0
 dtb-$(CONFIG_ARCH_CODASIP) += a70x-hobgoblin-qemu.dtb
 dtb-$(CONFIG_ARCH_CODASIP) += a70x-hobgoblin-etherlite.dtb
+dtb-$(CONFIG_ARCH_CODASIP) += a70x-hobgoblin-axi-emac.dtb
diff --git a/arch/riscv/boot/dts/codasip/a70x-hobgoblin-axi-emac.dts b/arch/riscv/boot/dts/codasip/a70x-hobgoblin-axi-emac.dts
new file mode 100644
index 000000000000..e7943823052c
--- /dev/null
+++ b/arch/riscv/boot/dts/codasip/a70x-hobgoblin-axi-emac.dts
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024 Codasip Limited
+ *
+ * A730 Hobgoblin FPGA platform, AXI ethernet block
+ */
+
+#include "a70x-hobgoblin.dts"
+
+&soc {
+	axi_ethernet: ethernet@0x600c0000 {
+		compatible = "xlnx,axi-ethernet-1.00.a";
+		interrupts-extended = <&plic 15>;
+		clock-names = "s_axi_lite_clk";
+		clocks = <&clk_100m>;
+		phy-mode = "rgmii";
+		reg = <0x600c0000 0x40000>;
+		xlnx,rxcsum = <2>;
+		xlnx,rxmem = <0x4000>;
+		xlnx,txcsum = <2>;
+		phy-handle = <&phy0>;
+		axistream-connected = <&axistream_dma>;
+
+		mdio {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			phy0: ethernet-phy@1 {
+				device_type = "ethernet-phy";
+				reg = <1>;
+				interrupts-extended = <&plic 18>;
+			};
+		};
+	};
+	/* U-boot does not support specifying the DMA directly in the
+	 * axi_ethernet above. The kernel does support it though
+	 */
+	axistream_dma: axistream_dma@0x600a0000  {
+		reg = <0x600a0000 0x10000>;
+		interrupts-extended = <&plic 16>, <&plic 17>;
+	};
+};
diff --git a/arch/riscv/boot/dts/codasip/a70x-hobgoblin.dts b/arch/riscv/boot/dts/codasip/a70x-hobgoblin.dts
index 7e6e3ea77be0..2333bb9f7e6c 100644
--- a/arch/riscv/boot/dts/codasip/a70x-hobgoblin.dts
+++ b/arch/riscv/boot/dts/codasip/a70x-hobgoblin.dts
@@ -70,6 +70,7 @@ soc: soc@0 {
 		compatible = "simple-bus";
 		/* child-bus-address parent-bus-address*2 length */
 		ranges = <0x0 0x0 0x0 0x80000000>;
+		dma-noncoherent;
 
 		plic: interrupt-controller@40000000 {
 			compatible = "codasip,plic", "riscv,plic0";
@@ -181,13 +182,13 @@ sdcard_hog {
 				line-name = "sdcard-fast-gpio";
 			};
 		};
-	};
 
-	clock: clock {
-		compatible = "fixed-clock";
-		#clock-cells = <0>;
-                clock-frequency = <20000>;
-		clock-accuracy = <100>;
+		/* Interconnect runs off this clock */
+		clk_100m: clock@100 {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <100000000>;
+		};
 	};
 
 	gpio-keys {
diff --git a/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig b/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig
index 5266d28ed64e..85581f990306 100644
--- a/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig
+++ b/arch/riscv/configs/codasip-a70x-hobgoblin_defconfig
@@ -57,6 +57,7 @@ CONFIG_VIRTIO_BLK=y
 CONFIG_NETDEVICES=y
 CONFIG_VIRTIO_NET=y
 CONFIG_XILINX_EMACLITE=y
+CONFIG_XILINX_AXI_EMAC=y
 CONFIG_REALTEK_PHY=y
 # CONFIG_WLAN is not set
 CONFIG_INPUT_MOUSEDEV=y
@@ -98,6 +99,8 @@ CONFIG_LEDS_TRIGGER_HEARTBEAT=y
 CONFIG_RTC_CLASS=y
 # CONFIG_RTC_NVMEM is not set
 CONFIG_RTC_DRV_GOLDFISH=y
+CONFIG_DMADEVICES=y
+CONFIG_XILINX_DMA=y
 CONFIG_SYNC_FILE=y
 CONFIG_VIRTIO_MMIO=y
 # CONFIG_VHOST_MENU is not set
-- 
2.39.3

